### PR TITLE
feat: fetch data and filter functionality

### PR DESCRIPTION
**Pull request by pankaj-prog**
feat: fetch data from mock backend
feat: filter functionality courses page

### CODE REVIEW

File: src/pages/Courses/utils/filterReducer.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+export const filterReducer = (state, action) => {
```
Comment (PuneetP16) [inline src/pages/Courses/utils/filterReducer.js:1]: Even if it was intentional, here's what I noticed:
1. Clear All button is not resetting all the filter states. 
2. When moving to another page and returning to the products page, the filter state is getting reset.
3. Couldn't find the values of the rating being rendered on the Product Card.

image

File: src/pages/Courses/utils/filterReducer.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+        ...state,
+        category: [
+          ...state.category.filter((category) => category !== action.payload),
+        ],
+      };
+    case "price_range":
+      return { ...state, minPrice: action.payload };
+    default:
+      throw new Error(`unknown action type:  ${action.type}`);
+  }
```
Comment (PuneetP16) [inline src/pages/Courses/utils/filterReducer.js:34]: Appreciation: Catching the dispatch error cases is what I will implement on my project too.

File: src/pages/Courses/utils/getFilteredData.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        return product2.discountedPrice - product1.discountedPrice;
+      }
+    }),
+  ];
+
+  // filter by rating
+
+  let filteredData = [
+    ...sortedProductList.filter((product) => product.rating >= minRating),
+  ];
```
Comment (PuneetP16) [inline src/pages/Courses/utils/getFilteredData.js]: Suggestion: filter method does give a brand new array, so no need to spread it deliberately

**Review by m26unkwn — APPROVED** [review]
Everything seems good and clean to me.
Things that I like the most.
1. using reducer for filter inside the product listing page
2. Make clean and readable function for filters


You can add clear filters too

File: src/pages/Courses/utils/getFilteredData.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+      ]
+    : [...filteredData];
+
+  // filter by price range
+
+  filteredData = [
+    ...filteredData.filter((product) => {
+      return Number(product.discountedPrice) > minPrice;
+    }),
+  ];
```
Comment (PuneetP16) [inline src/pages/Courses/utils/getFilteredData.js]: Suggestion: You could strip the code down and still have same effect:
filteredData = filteredData.filter((product) =>Number(product.discountedPrice) > minPrice);

**Review by PuneetP16 — APPROVED** [review]
Left few suggestions to work upon. Did great job on building the clean logic.

**Pull request merged**